[AutoDiff] Unify code for differentiation parameter inference.

dan-zheng opened the pull request:
Previously, `@differentiable` attribute type-checking used ad-hoc
differentiation parameter inference logic.

Now, that differentiation parameter inference logic has been moved to
`AutoDiffParameterIndicesBuilder::inferParameters` so that it can be shared
with `@differentiable` attribute type-checking.

Add retroactive derivative tests.

Resolves [TF-363](https://bugs.swift.org/browse/TF-363).

dan-zheng committed:
[AutoDiff] Unify code for differentiation parameter inference.

Previously, `@differentiable` attribute type-checking used ad-hoc
differentiation parameter inference logic.

Now, that differentiation parameter inference logic has been moved to
`AutoDiffParameterIndicesBuilder::inferParameters` so that it can be shared
with `@differentiable` attribute type-checking.

Add retroactive derivative tests.

Resolves TF-363.

dan-zheng commented on include/swift/AST/AutoDiff.h:
```diff
@@ -202,9 +206,6 @@ class AutoDiffParameterIndicesBuilder {
   void setParameters(unsigned lowerBound, unsigned upperBound);
 
-  /// Sets all parameters.
-  void setAllParameters();
```
Removed `setAllParameters` because it has no usages and is subsumed by `inferParameters`.

It would be possible to implement `inferParameters` using `setAllParameters` if `AutoDiffParameterIndicesBuilder` stored a `functionType`, but it doesn't.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented on lib/AST/AutoDiff.cpp:
```diff
@@ -261,9 +262,60 @@ static unsigned getNumAutoDiffParameterIndices(AnyFunctionType *fnTy) {
 
+/// Returns true if the given type conforms to `Differentiable` in the given
+/// module.
+static bool conformsToDifferentiableInModule(Type type, ModuleDecl *module) {
```
Note: `conformsToDifferentiableInModule` is duped in `TypeCheckAttr.cpp`. It could be declared in `AutoDiff.h` to eliminate dupe, holding off for now though.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on include/swift/AST/AutoDiff.h:
```diff
@@ -187,9 +187,13 @@ class AutoDiffParameterIndicesBuilder {
+
+  /// Creates a `AutoDiffParameterIndicesBuilder` for the given function type,
+  /// inferring all differentiation parameters.
+  static AutoDiffParameterIndicesBuilder inferParameters(
```
It would be useful to document the inference rules here.

dan-zheng replied:
Inference rules documented in https://github.com/apple/swift/pull/23350/commits/487fc215654af3737f7dcaefd9c6c95224d9a554.

rxwei commented on test/AutoDiff/retroactive_derivatives.swift:
```diff
@@ -0,0 +1,63 @@
+
+import StdlibUnittest
+
+var RetroactiveDerivativeTests = TestSuite("RetroactiveDerivative")
```
* Retroactive derivatives will be the only way to register derivatives in the future. Thus this test file name and test suite name should ideally be general like "derivative_registration.swift".
* There's a `@differentiating` test in `custom_derivatives.swift`. Do you want to move it here? We could also merge all derivative registration tests into `custom_derivatives.swift`.

dan-zheng replied:
I moved all `@differentiating` tests to `test/AutoDiff/derivative_registration.swift`.

I chose not to move tests into `custom_derivatives.swift` because that file also contains tests for other APIs (`withGradient`, `withRecomputationInPullbacks`, etc).

dan-zheng committed:
Address review comments.

- Move tests to `test/AutoDiff/derivative_registration.swift`.
- Document differentiation parameter inference rules.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow macOS

dan-zheng commented:
@swift-ci Please test tensorflow linux GPU
